### PR TITLE
Support builtin and numpy types in HDF5

### LINKED ISSUE #3056: DataGroup HDF5 I/O with non-scipp items
Currently we do not support plain Python types or NumPy arrays. Fully generic support may be problematic, but we could support, e.g., `numpy.ndarray`, `str`, `int`, and `float`.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #3056

### CODE REVIEW

**Comment by jl-wynen** [comment]
How do we handle the version of the file format here? Files with these entries cannot be loaded by an old scipp version. But old files continue to work.

File: tests/io/hdf5_test.py [hunk 1]
```diff
@@ -279,6 +279,18 @@ def test_data_group():
...
+            'numpy int64': np.int64(9),
+            'numpy int32': np.int32(-23),
+            'numpy uint64': np.uint64(9312),
+            'numpy uint32': np.uint32(66134),
+            'float': 7.81,
+            'numpy float64': np.float64(5235.1),
+            'numpy float32': np.float32(0.031),
+            'boolean': True,
+            'numpy bool_': np.bool_(False),
+            'string': 'a string for testing',
```
Comment (SimonHeybrock) [inline tests/io/hdf5_test.py:292]: I have seen NeXus files that use bytesrings, can we support that?

**Comment by SimonHeybrock** [comment]
> How do we handle the version of the file format here? Files with these entries cannot be loaded by an old scipp version. But old files continue to work.

We have so far not done any version checks on files, have we?

File: tests/io/hdf5_test.py (hunk 1 above)
Comment (jl-wynen) [reply tests/io/hdf5_test.py:292]: What do you mean by bytestring? Python's `bytes`? h5py seems to store `str` as `bytes`. This is why the loader calls `decode`.

**Comment by jl-wynen** [comment]
> > How do we handle the version of the file format here? Files with these entries cannot be loaded by an old scipp version. But old files continue to work.
> 
> We have so far not done any version checks on files, have we?

We encode a version in the file but we don't actually check it for some reason.

File: tests/io/hdf5_test.py (hunk 1 above)
Comment (SimonHeybrock) [reply tests/io/hdf5_test.py:292]: Yes, `bytes`, iirc.

**Comment by SimonHeybrock** [comment]
> > > How do we handle the version of the file format here? Files with these entries cannot be loaded by an old scipp version. But old files continue to work.
> > 
> > 
> > We have so far not done any version checks on files, have we?
> 
> We encode a version in the file but we don't actually check it for some reason.

It was added as a way of keeping that option open in the future. Maybe we should?

File: tests/io/hdf5_test.py (hunk 1 above)
Comment (jl-wynen) [reply tests/io/hdf5_test.py:292]: Done

**Comment by jl-wynen** [comment]
> >    We encode a version in the file but we don't actually check it for some reason.

> It was added as a way of keeping that option open in the future. Maybe we should?

Unfortunately, it stores the scipp version not a version of the format. As far as I can tell, when loading a file with an old version, there is no way to check if the file comes from a new, incompatible version of scipp.
If we want such a feature, we should implement support separately from this PR.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**